src/core/expression/qgsexpressionutils.cpp

```diff
@@ -136,26 +181,103 @@ void QgsExpressionUtils::executeLambdaForMapLayer( const QVariant &value, const
     return;
   }
 
-  // if no layer stores, then this is only for layers in project and therefore associated with the main thread
-  auto runFunction = [ value, context, expression, &function, &foundLayer ]
+  if ( !context || context->layerStores().empty() )
   {
-    if ( QgsMapLayer *layer = getMapLayerPrivate( value, context, expression ) )
+    // if no layer stores, then this is only for layers in project and therefore associated with the main thread
+    auto runFunction = [ value, context, expression, &function, &foundLayer ]
     {
-      foundLayer = true;
-      function( layer );
-    }
+      if ( QgsMapLayer *layer = getMapLayerPrivate( value, context, expression ) )
+      {
+        foundLayer = true;
+        function( layer );
+      }
+      else
+      {
+        foundLayer = false;
+      }
+    };
+
+    // Make sure we only deal with the project on the thread where it lives.
+    // Anything else risks a crash.
+    if ( QThread::currentThread() == QgsProject::instance()->thread() )
+      runFunction();
     else
+      QMetaObject::invokeMethod( QgsProject::instance(), runFunction, Qt::BlockingQueuedConnection );
+  }
+  else
+  {
+    // if layer stores, then we can't be certain in advance of which thread the layer will have affinity with.
+    // So we need to fetch the layer and then run the function on the layer's thread.
+
+    const QString identifier = value.toString();
+    // check through layer stores from context
+    if ( context )
```

Comment:
You don't need to repeat the pointer null check here as it's done above (https://github.com/qgis/QGIS/pull/51192/files#diff-1f7dbcbc646b53b8e65fb537946c39decebe9262170e958cc8c5bd42260de7b5R184)

Reply:
thanks, fixed